File: hutulock-client/src/main/java/com/hutulock/client/HutuLockClient.java
/*
 * Copyright 2026 HutuLock Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hutulock.client;

import com.hutulock.config.api.ClientProperties;
import com.hutulock.model.protocol.CommandType;
import com.hutulock.model.protocol.Message;
import com.hutulock.model.watcher.WatchEvent;
import io.netty.bootstrap.Bootstrap;
import io.netty.channel.*;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioSocketChannel;
import io.netty.handler.codec.LineBasedFrameDecoder;
import io.netty.handler.codec.string.StringDecoder;
import io.netty.handler.codec.string.StringEncoder;
import io.netty.util.CharsetUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.*;

/**
 * HutuLock 分布式锁客户端
 *
 * <p>核心设计（ZooKeeper 风格）：
 * <ol>
 *   <li>Session 会话：独立于 TCP 连接，断线重连后可恢复</li>
 *   <li>Watcher 驱动：收到 NODE_DELETED 事件后重新检查锁，而不是服务端主动推送</li>
 *   <li>顺序临时节点：公平锁，每个等待者只监听前一个节点，避免羊群效应</li>
 *   <li>看门狗：持锁期间定时发送 RENEW 心跳，防止服务端 TTL 过期</li>
 * </ol>
 *
 * <p>简单用法：
 * <pre>{@code
 *   HutuLockClient client = HutuLockClient.builder()
 *       .addNode("127.0.0.1", 8881)
 *       .addNode("127.0.0.1", 8882)
 *       .build();
 *   client.connect();
 *
 *   boolean held = client.lock("order-lock");
 *   try { ... } finally { client.unlock("order-lock"); }
 * }</pre>
 *
 * <p>带看门狗的完整用法：
 * <pre>{@code
 *   LockContext ctx = LockContext.builder("order-lock", client.getSessionId())
 *       .ttl(30, TimeUnit.SECONDS)
 *       .watchdogInterval(10, TimeUnit.SECONDS)
 *       .onExpired(name -> abortWork())
 *       .build();
 *   client.lock(ctx);
 *   try { ... } finally { client.unlock(ctx); }
 * }</pre>
 *
 * @author HutuLock Authors
 * @since 1.0.0
 */
public class HutuLockClient implements AutoCloseable {

    private static final Logger log = LoggerFactory.getLogger(HutuLockClient.class);

    private final ClientProperties config;
    private final List<String[]>   nodes = new ArrayList<>();

    private volatile Channel           channel;
    private volatile LockClientHandler handler;
    private volatile String            sessionId;

    private final EventLoopGroup group = new NioEventLoopGroup();

    /** lockName → LockContext（持有的锁） */
    private final Map<String, LockContext> heldContexts = new ConcurrentHashMap<>();

    /** 共享看门狗调度器，所有 LockContext 复用 */
    private final ScheduledExecutorService watchdogScheduler =
        Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "hutulock-watchdog");
            t.setDaemon(true);
            return t;
        });

    private HutuLockClient(Builder b) {
        this.config = b.config;
        this.nodes.addAll(b.nodes);
    }

    public String getSessionId() { return sessionId; }

    // ==================== 连接与会话 ====================

    /**
     * 连接到集群并建立会话。
     *
     * @throws Exception 所有节点均不可达
     */
    public void connect() throws Exception {
        connectToAny(new ArrayList<>(nodes));
        this.sessionId = establishSession(null);
        log.info("Connected, sessionId={}", sessionId);
    }

    private String establishSession(String existingSessionId) throws Exception {
        String arg = existingSessionId != null ? existingSessionId : "";
        Message req = arg.isEmpty()
            ? Message.of(CommandType.CONNECT)
            : Message.of(CommandType.CONNECT, arg);

        CompletableFuture<Message> future = handler.registerRequest("CONNECT");
        channel.writeAndFlush(req.serialize() + "\n");

        Message resp = future.get(config.connectTimeoutMs, TimeUnit.MILLISECONDS);
        if (resp.getType() != CommandType.CONNECTED) {
            throw new RuntimeException("Failed to establish session: " + resp);
        }
        return resp.arg(0);
    }

    private void connectToAny(List<String[]> candidates) throws Exception {
        Exception last = null;
        for (String[] node : candidates) {
            try {
                doConnect(node[0], Integer.parseInt(node[1]));
                return;
            } catch (Exception e) {
                last = e;
                log.warn("Failed to connect to {}:{}", node[0], node[1]);
            }
        }
        throw new RuntimeException("All nodes unreachable", last);
    }

    private void doConnect(String host, int port) throws InterruptedException {
        LockClientHandler newHandler = new LockClientHandler();
        newHandler.setRedirectListener(this::handleRedirect);

        channel = new Bootstrap()
            .group(group)
            .channel(NioSocketChannel.class)
            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, config.connectTimeoutMs)
            .handler(new ChannelInitializer<SocketChannel>() {
                @Override
                protected void initChannel(SocketChannel ch) {
                    ch.pipeline()
                      .addLast(new LineBasedFrameDecoder(config.maxFrameLength))
                      .addLast(new StringDecoder(CharsetUtil.UTF_8))
                      .addLast(new StringEncoder(CharsetUtil.UTF_8))
                      .addLast(newHandler);
                }
            })
            .connect(host, port).sync().channel();
        handler = newHandler;
        log.info("Connected to {}:{}", host, port);
    }

    // ==================== 简单 API ====================

    /** 使用默认配置获取锁（自动创建 LockContext）。 */
    public boolean lock(String lockName) throws Exception {
        LockContext ctx = LockContext.builder(lockName, sessionId)
            .scheduler(watchdogScheduler)
            .build();
        return lock(ctx);
    }

    /** 使用默认配置释放锁。 */
    public void unlock(String lockName) throws Exception {
        LockContext ctx = heldContexts.get(lockName);
        if (ctx == null) throw new IllegalStateException("lock [" + lockName + "] not held");
        unlock(ctx);
    }

    // ==================== Context API ====================

    /**
     * 使用 LockContext 获取锁（ZooKeeper 顺序节点 + Watcher 驱动）。
     *
     * <p>流程：
     * <ol>
     *   <li>发送 LOCK → 服务端创建 EPHEMERAL_SEQ 节点</li>
     *   <li>收到 OK → 获锁成功，启动看门狗</li>
     *   <li>收到 WAIT seqNodePath → 等待 NODE_DELETED Watcher 事件</li>
     *   <li>收到 WATCH_EVENT NODE_DELETED → 发送 RECHECK → 重新检查</li>
     * </ol>
     */
    public boolean lock(LockContext ctx) throws Exception {
        return lock(ctx, config.lockTimeoutS, TimeUnit.SECONDS);
    }

    public boolean lock(LockContext ctx, long timeout, TimeUnit unit) throws Exception {
        long deadline = System.currentTimeMillis() + unit.toMillis(timeout);

        while (System.currentTimeMillis() < deadline) {
            CompletableFuture<Message> future = handler.registerRequest("LOCK:" + ctx.getLockName());
            channel.writeAndFlush(
                Message.of(CommandType.LOCK, ctx.getLockName(), sessionId).serialize() + "\n");

            long remaining = deadline - System.currentTimeMillis();
            Message resp = future.get(Math.max(remaining, 1000), TimeUnit.MILLISECONDS);

            if (resp.getType() == CommandType.REDIRECT) {
                handleRedirect(resp.arg(0));
                continue;
            }

            if (resp.getType() == CommandType.OK) {
                ctx.setSeqNodePath(resp.arg(1));
                ctx.startWatchdog(channel);
                heldContexts.put(ctx.getLockName(), ctx);
                log.info("Lock acquired: lock={}, seq={}", ctx.getLockName(), ctx.getSeqNodePath());
                return true;
            }

            if (resp.getType() == CommandType.WAIT) {
                ctx.setSeqNodePath(resp.arg(1));
                if (waitForLock(ctx, deadline)) return true;
                break;
            }
        }
        return false;
    }

    /**
     * 等待 Watcher 事件，收到后重新检查锁（ZooKeeper 核心模式）。
     */
    private boolean waitForLock(LockContext ctx, long deadline) throws Exception {
        CompletableFuture<WatchEvent> watchFuture = new CompletableFuture<>();
        handler.registerWatcher(ctx.getSeqNodePath(), watchFuture::complete);

        long remaining = deadline - System.currentTimeMillis();
        if (remaining <= 0) return false;

        WatchEvent event = watchFuture.get(remaining, TimeUnit.MILLISECONDS);
        if (event.getType() == WatchEvent.Type.SESSION_EXPIRED) {
            ctx.markExpired();
            return false;
        }

        return recheckLock(ctx, deadline);
    }

    /**
     * 重新检查锁（Watcher 触发后重新检查锁）。
     */
    private boolean recheckLock(LockContext ctx, long deadline) throws Exception {
        CompletableFuture<Message> future =
            handler.registerRequest("RECHECK:" + ctx.getSeqNodePath());
        channel.writeAndFlush(
            Message.of(CommandType.RECHECK, ctx.getLockName(), ctx.getSeqNodePath(), sessionId)
                   .serialize() + "\n");

        long remaining = deadline - System.currentTimeMillis();
        Message resp = future.get(Math.max(remaining, 1000), TimeUnit.MILLISECONDS);

        if (resp.getType() == CommandType.OK) {
            ctx.startWatchdog(channel);
            heldContexts.put(ctx.getLockName(), ctx);
            log.info("Lock acquired after recheck: {}", ctx.getLockName());
            return true;
        }

        if (resp.getType() == CommandType.WAIT) {
            return waitForLock(ctx, deadline);
        }
        return false;
    }

    /**
     * 释放锁，停止看门狗。
     */
    public void unlock(LockContext ctx) throws Exception {
        ctx.stopWatchdog();
        heldContexts.remove(ctx.getLockName());

        CompletableFuture<Message> future =
            handler.registerRequest("UNLOCK:" + ctx.getLockName());
        channel.writeAndFlush(
            Message.of(CommandType.UNLOCK, ctx.getSeqNodePath(), sessionId).serialize() + "\n");

        try {
            future.get(10, TimeUnit.SECONDS);
            ctx.markReleased();
            log.info("Lock released: {}", ctx.getLockName());
        } catch (TimeoutException e) {
            log.warn("Unlock timeout for [{}], session expiry will clean up", ctx.getLockName());
        }
    }

    // ==================== Optimistic locking API ====================

    /**
     * Read data and version from a ZNode path.
     *
     * <p>Use this to implement optimistic locking as a replacement for
     * MySQL {@code SELECT data, version FROM t WHERE id = ?}.
     *
     * @param path ZNode path, e.g. {@code /resources/order-123}
     * @return versioned data, or null if the node does not exist
     */
    public VersionedData getData(String path) throws Exception {
        CompletableFuture<Message> future = handler.registerRequest("GET_DATA:" + path);
        channel.writeAndFlush(
            Message.of(CommandType.GET_DATA, path, sessionId).serialize() + "\n");

        Message resp = future.get(10, TimeUnit.SECONDS);
        if (resp.getType() == CommandType.ERROR) return null;

        // Response: DATA <path> <version> <base64-data>
        int version = Integer.parseInt(resp.arg(1));
        byte[] data = resp.argCount() > 2
            ? java.util.Base64.getDecoder().decode(resp.arg(2))
            : new byte[0];
        return new VersionedData(path, data, version);
    }

    /**
     * Write data to a ZNode with optimistic version check.
     *
     * <p>Replaces MySQL {@code UPDATE t SET data=? WHERE id=? AND version=?}.
     *
     * @param path    ZNode path
     * @param data    new data
     * @param version expected current version (from a prior {@link #getData} call)
     * @return true if write succeeded, false if version mismatch (retry needed)
     */
    public boolean setData(String path, byte[] data, int version) throws Exception {
        String encoded = java.util.Base64.getEncoder().encodeToString(data);
        CompletableFuture<Message> future = handler.registerRequest("SET_DATA:" + path);
        channel.writeAndFlush(
            Message.of(CommandType.SET_DATA, path, String.valueOf(version), encoded, sessionId)
                   .serialize() + "\n");

        Message resp = future.get(10, TimeUnit.SECONDS);
        return resp.getType() == CommandType.OK;
    }

    /** Convenience overload for String data. */
    public boolean setData(String path, String data, int version) throws Exception {
        return setData(path, data.getBytes(java.nio.charset.StandardCharsets.UTF_8), version);
    }

    /**
     * Optimistic update with automatic retry.
     *
     * <p>Reads current data, applies {@code updater}, writes back with version check.
     * Retries up to {@code maxRetries} times on VERSION_MISMATCH.
     *
     * <pre>{@code
     *   // Replace MySQL optimistic lock pattern:
     *   boolean ok = client.optimisticUpdate("/resources/order-123", maxRetries, current -> {
     *       Order order = deserialize(current.getData());
     *       order.setStatus("PAID");
     *       return serialize(order);
     *   });
     * }</pre>
     *
     * @param path       ZNode path
     * @param maxRetries max retry attempts on version conflict
     * @param updater    function that receives current data and returns new data
     * @return true if update succeeded within maxRetries attempts
     */
    public boolean optimisticUpdate(String path, int maxRetries,
                                    java.util.function.Function<VersionedData, byte[]> updater)
            throws Exception {
        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            VersionedData current = getData(path);
            if (current == null) return false;

            byte[] newData = updater.apply(current);
            if (setData(path, newData, current.getVersion())) return true;

            if (attempt < maxRetries) {
                log.debug("Version conflict on {}, retry {}/{}", path, attempt + 1, maxRetries);
                Thread.sleep(10L * (1 << Math.min(attempt, 5))); // exponential backoff
            }
        }
        log.warn("optimisticUpdate failed after {} retries on {}", maxRetries, path);
        return false;
    }

    // ==================== 重连 ====================

    private void handleRedirect(String leaderId) {
        log.info("Redirected to leader: {}, reconnecting...", leaderId);
        try {
            if (channel != null) channel.close().sync();
            List<String[]> shuffled = new ArrayList<>(nodes);
            Collections.shuffle(shuffled);
            connectToAny(shuffled);
            this.sessionId = establishSession(this.sessionId);
        } catch (Exception e) {
            log.error("Reconnect failed", e);
        }
    }

    // ==================== 关闭 ====================

    @Override
    public void close() {
        heldContexts.values().forEach(LockContext::stopWatchdog);
        watchdogScheduler.shutdownNow();
        if (channel != null) channel.close();
        group.shutdownGracefully();
    }

    // ==================== Builder ====================

    public static Builder builder() { return new Builder(); }

    /** HutuLockClient 构建器 */
    public static final class Builder {
        private ClientProperties config = ClientProperties.defaults();
        private final List<String[]> nodes = new ArrayList<>();

        public Builder config(ClientProperties config) { this.config = config; return this; }

        public Builder addNode(String host, int port) {
            nodes.add(new String[]{host, String.valueOf(port)});
            return this;
        }

        public HutuLockClient build() {
            if (nodes.isEmpty()) throw new IllegalStateException("at least one node required");
            return new HutuLockClient(this);
        }
    }
}

File: sdk/php/src/HutuLockClient.php
<?php

declare(strict_types=1);

namespace HutuLock;

/**
 * HutuLock 分布式锁 PHP 客户端
 *
 * 协议：纯文本行（UTF-8），与 Java/Python/Go 客户端完全兼容。
 *
 * 用法：
 *   $client = new HutuLockClient([['127.0.0.1', 8881]]);
 *   $client->connect();
 *
 *   $token = $client->lock('order-lock');
 *   try {
 *       // 临界区
 *   } finally {
 *       $client->unlock($token);
 *   }
 *   $client->close();
 *
 * 要求：PHP 8.0+，无第三方依赖（仅 sockets 扩展）。
 */
class HutuLockClient
{
    /** @var array<array{0:string,1:int}> */
    private array $nodes;
    private float $connectTimeout;
    private float $lockTimeout;
    private float $watchdogInterval;

    /** @var resource|null TCP socket */
    private $socket = null;
    private ?string $sessionId = null;

    /** @var array<string, array{resolve: callable, reject: callable}> key => pending request */
    private array $pending = [];
    /** @var array<string, callable> seqPath => watcher callback */
    private array $watchers = [];

    /** @var string 未处理完的读缓冲区 */
    private string $readBuf = '';

    /** @var array<string, LockToken> lockName => token */
    private array $held = [];

    /** @var array<string, bool> lockName => watchdog running */
    private array $watchdogRunning = [];

    /**
     * @param array<array{0:string,1:int}> $nodes  集群节点列表，格式 [['host', port], ...]
     * @param float $connectTimeout  连接超时（秒），默认 5
     * @param float $lockTimeout     获取锁超时（秒），默认 30
     * @param float $watchdogInterval 看门狗心跳间隔（秒），默认 10
     */
    public function __construct(
        array $nodes,
        float $connectTimeout  = 5.0,
        float $lockTimeout     = 30.0,
        float $watchdogInterval = 10.0
    ) {
        if (empty($nodes)) {
            throw new \InvalidArgumentException('at least one node required');
        }
        $this->nodes            = $nodes;
        $this->connectTimeout   = $connectTimeout;
        $this->lockTimeout      = $lockTimeout;
        $this->watchdogInterval = $watchdogInterval;
    }

    public function getSessionId(): ?string
    {
        return $this->sessionId;
    }

    // ---- 连接 ----

    public function connect(): void
    {
        $nodes = $this->nodes;
        shuffle($nodes);
        $lastErr = null;

        foreach ($nodes as [$host, $port]) {
            try {
                $this->doConnect($host, $port);
                $this->sessionId = $this->establishSession();
                return;
            } catch (\Throwable $e) {
                $lastErr = $e;
            }
        }
        throw new ConnectionException('all nodes unreachable: ' . ($lastErr?->getMessage() ?? ''));
    }

    private function doConnect(string $host, int $port): void
    {
        $errno  = 0;
        $errstr = '';
        $sock = @fsockopen($host, $port, $errno, $errstr, $this->connectTimeout);
        if ($sock === false) {
            throw new ConnectionException("connect to {$host}:{$port} failed: {$errstr} ({$errno})");
        }
        stream_set_timeout($sock, (int)$this->connectTimeout, (int)(($this->connectTimeout - floor($this->connectTimeout)) * 1_000_000));
        stream_set_blocking($sock, false);
        $this->socket  = $sock;
        $this->readBuf = '';
    }

    private function establishSession(?string $existingId = null): string
    {
        $args = $existingId !== null ? [$existingId] : [];
        $resp = $this->request('CONNECT', new Message(Message::CONNECT, $args), $this->connectTimeout);
        if ($resp->type !== Message::CONNECTED) {
            throw new ProtocolException('unexpected response to CONNECT: ' . $resp->serialize());
        }
        return $resp->arg(0);
    }

    // ---- 获取锁 ----

    /**
     * 获取锁，返回 LockToken。超时抛出 LockTimeoutException。
     */
    public function lock(string $lockName, ?float $timeout = null): LockToken
    {
        $deadline = microtime(true) + ($timeout ?? $this->lockTimeout);
        $token    = $this->doAcquire($lockName, $deadline);
        $this->startWatchdog($token);
        $this->held[$lockName] = $token;
        return $token;
    }

    private function doAcquire(string $lockName, float $deadline): LockToken
    {
        $remaining = $deadline - microtime(true);
        if ($remaining <= 0) {
            throw new LockTimeoutException($lockName);
        }

        $resp = $this->request(
            "LOCK:{$lockName}",
            new Message(Message::LOCK, [$lockName, $this->sessionId]),
            $remaining
        );

        if ($resp->type === Message::REDIRECT) {
            $this->handleRedirect($resp->arg(0));
            return $this->doAcquire($lockName, $deadline);
        }

        if ($resp->type === Message::OK) {
            return new LockToken($lockName, $resp->arg(1), $this->sessionId);
        }

        if ($resp->type === Message::WAIT) {
            return $this->waitForLock($lockName, $resp->arg(1), $deadline);
        }

        throw new ProtocolException('unexpected response to LOCK: ' . $resp->serialize());
    }

    private function waitForLock(string $lockName, string $seqPath, float $deadline): LockToken
    {
        $remaining = $deadline - microtime(true);
        if ($remaining <= 0) {
            throw new LockTimeoutException($lockName);
        }

        $eventType = null;
        $this->watchers[$seqPath] = function (string $type) use (&$eventType): void {
            $eventType = $type;
        };

        // 轮询读取，等待 WATCH_EVENT
        $waitUntil = microtime(true) + $remaining;
        while (microtime(true) < $waitUntil) {
            $this->readAndDispatch(0.05);
            if ($eventType !== null) {
                break;
            }
        }

        if ($eventType === null) {
            unset($this->watchers[$seqPath]);
            throw new LockTimeoutException($lockName);
        }
        if ($eventType === 'SESSION_EXPIRED') {
            throw new SessionExpiredException();
        }

        return $this->recheck($lockName, $seqPath, $deadline);
    }

    private function recheck(string $lockName, string $seqPath, float $deadline): LockToken
    {
        $remaining = $deadline - microtime(true);
        if ($remaining <= 0) {
            throw new LockTimeoutException($lockName);
        }

        $resp = $this->request(
            "RECHECK:{$seqPath}",
            new Message(Message::RECHECK, [$lockName, $seqPath, $this->sessionId]),
            $remaining
        );

        if ($resp->type === Message::OK) {
            return new LockToken($lockName, $resp->arg(1), $this->sessionId);
        }
        if ($resp->type === Message::WAIT) {
            return $this->waitForLock($lockName, $resp->arg(1), $deadline);
        }
        throw new ProtocolException('unexpected response to RECHECK: ' . $resp->serialize());
    }

    // ---- 释放锁 ----

    public function unlock(LockToken $token): void
    {
        $token->stopWatchdog();
        unset($this->held[$token->lockName], $this->watchdogRunning[$token->lockName]);

        try {
            $this->request(
                "UNLOCK:{$token->lockName}",
                new Message(Message::UNLOCK, [$token->seqNodePath, $this->sessionId]),
                10.0
            );
        } catch (LockTimeoutException $e) {
            // 超时静默处理，服务端 TTL 会清理
        }
    }

    // ---- 看门狗（同步轮询实现） ----

    /**
     * PHP 单线程无法后台定时，看门狗通过 tick 函数模拟。
     * 调用 lock() 后，每次调用 tick() 会检查是否需要续期。
     * 也可在业务循环中定期调用 tick()。
     */
    private array $watchdogNextRenew = [];

    private function startWatchdog(LockToken $token): void
    {
        $this->watchdogRunning[$token->lockName]  = true;
        $this->watchdogNextRenew[$token->lockName] = microtime(true) + $this->watchdogInterval;

        $token->setStopWatchdog(function () use ($token): void {
            unset(
                $this->watchdogRunning[$token->lockName],
                $this->watchdogNextRenew[$token->lockName]
            );
        });
    }

    /**
     * 驱动看门狗续期，在持锁期间的业务循环中调用。
     * 若不调用，看门狗不会自动续期（PHP 单线程限制）。
     */
    public function tick(): void
    {
        $now = microtime(true);
        foreach ($this->watchdogRunning as $lockName => $_) {
            if (isset($this->watchdogNextRenew[$lockName]) && $now >= $this->watchdogNextRenew[$lockName]) {
                try {
                    $this->request(
                        "RENEW:{$lockName}",
                        new Message(Message::RENEW, [$lockName, $this->sessionId]),
                        5.0
                    );
                } catch (\Throwable $e) {
                    // 续期失败不抛出，下次再试
                }
                $this->watchdogNextRenew[$lockName] = $now + $this->watchdogInterval;
            }
        }
    }

    // ---- 重定向 ----

    private function handleRedirect(string $leaderId): void
    {
        if ($this->socket !== null) {
            fclose($this->socket);
            $this->socket = null;
        }
        $nodes = $this->nodes;
        shuffle($nodes);
        $lastErr = null;
        foreach ($nodes as [$host, $port]) {
            try {
                $this->doConnect($host, $port);
                $this->sessionId = $this->establishSession($this->sessionId);
                return;
            } catch (\Throwable $e) {
                $lastErr = $e;
            }
        }
        throw new ConnectionException('reconnect failed: ' . ($lastErr?->getMessage() ?? ''));
    }

    // ---- 底层 request/dispatch ----

    /**
     * 发送消息并同步等待对应 key 的响应。
     */
    private function request(string $key, Message $msg, float $timeout): Message
    {
        $this->send($msg);

        $result   = null;
        $error    = null;
        $deadline = microtime(true) + $timeout;

        $this->pending[$key] = [
            'resolve' => function (Message $m) use (&$result): void { $result = $m; },
            'reject'  => function (\Throwable $e) use (&$error): void { $error = $e; },
        ];

        while ($result === null && $error === null && microtime(true) < $deadline) {
            $this->readAndDispatch(min(0.05, $deadline - microtime(true)));
        }

        unset($this->pending[$key]);

        if ($error !== null) {
            throw $error;
        }
        if ($result === null) {
            throw new LockTimeoutException($key);
        }
        return $result;
    }

    private function send(Message $msg): void
    {
        if ($this->socket === null) {
            throw new ConnectionException('not connected');
        }
        $line = $msg->serialize() . "\n";
        $written = fwrite($this->socket, $line);
        if ($written === false) {
            throw new ConnectionException('write failed');
        }
    }

    /**
     * 非阻塞读取并分发响应，最多等待 $maxWait 秒。
     */
    private function readAndDispatch(float $maxWait): void
    {
        if ($this->socket === null) {
            return;
        }

        $read   = [$this->socket];
        $write  = null;
        $except = null;
        $sec    = (int)$maxWait;
        $usec   = (int)(($maxWait - $sec) * 1_000_000);

        $n = stream_select($read, $write, $except, $sec, $usec);
        if ($n === false || $n === 0) {
            return;
        }

        $chunk = fread($this->socket, 8192);
        if ($chunk === false || $chunk === '') {
            return;
        }
        $this->readBuf .= $chunk;

        while (($pos = strpos($this->readBuf, "\n")) !== false) {
            $line          = substr($this->readBuf, 0, $pos);
            $this->readBuf = substr($this->readBuf, $pos + 1);
            $this->dispatch(trim($line));
        }
    }

    private function dispatch(string $line): void
    {
        if ($line === '') {
            return;
        }

        // WATCH_EVENT 单独处理
        if (str_starts_with($line, Message::WATCH_EVENT)) {
            $this->handleWatchEvent($line);
            return;
        }

        try {
            $msg = Message::parse($line);
        } catch (\Throwable $e) {
            return;
        }

        switch ($msg->type) {
            case Message::CONNECTED:
                $this->complete('CONNECT', $msg);
                break;
            case Message::OK:
                $lockName = $msg->arg(0);
                $seqPath  = $msg->arg(1);
                $this->complete("LOCK:{$lockName}", $msg);
                $this->complete("RECHECK:{$seqPath}", $msg);
                break;
            case Message::WAIT:
                $this->complete('LOCK:' . $msg->arg(0), $msg);
                break;
            case Message::RELEASED:
                $this->complete('UNLOCK:' . $msg->arg(0), $msg);
                break;
            case Message::RENEWED:
                $this->complete('RENEW:' . $msg->arg(0), $msg);
                break;
            case Message::REDIRECT:
                $leaderId = $msg->arg(0);
                $err = new NotLeaderException($leaderId);
                foreach ($this->pending as $key => $cb) {
                    ($cb['reject'])($err);
                }
                $this->pending = [];
                break;
            case Message::ERROR:
                $errMsg = $msg->arg(0) ?: 'server error';
                $err = new ProtocolException($errMsg);
                foreach ($this->pending as $key => $cb) {
                    ($cb['reject'])($err);
                }
                $this->pending = [];
                break;
        }
    }

    private function handleWatchEvent(string $line): void
    {
        // 格式：WATCH_EVENT {type} {path}
        $parts = preg_split('/\s+/', $line, 3);
        if (count($parts) < 3) {
            return;
        }
        [, $eventType, $path] = $parts;
        if (isset($this->watchers[$path])) {
            $cb = $this->watchers[$path];
            unset($this->watchers[$path]);
            $cb($eventType);
        }
    }

    private function complete(string $key, Message $msg): void
    {
        if (isset($this->pending[$key])) {
            ($this->pending[$key]['resolve'])($msg);
            unset($this->pending[$key]);
        }
    }

    // ---- 关闭 ----

    public function close(): void
    {
        foreach ($this->held as $token) {
            $token->stopWatchdog();
        }
        $this->held    = [];
        $this->pending = [];
        $this->watchers = [];
        if ($this->socket !== null) {
            fclose($this->socket);
            $this->socket = null;
        }
    }

    public function __destruct()
    {
        $this->close();
    }
}
